packages/lib/src/actionMiddlewares/draftMiddleware.ts

```diff
@@ -0,0 +1,102 @@
+import { getParentToChildPath, resolvePath } from "../parent"
+import { applyPatches, onPatches, OnPatchesDisposer, patchRecorder } from "../patch"
+import { isRootStore, registerRootStore, unregisterRootStore } from "../rootStore"
+import { clone } from "../snapshot"
+import { assertTweakedObject } from "../tweaker/core"
+
+/**
+ * Callback function for `DraftManager.withDraft`.
+ */
+export type WithDraftCallback<T extends object> = (draftNode: T) => boolean | undefined | void
+
+/**
+ * Manager class returned by `draftMiddleware` that allows you to make changes to a draft copy of
+ * the original subtree and apply them to the original subtree or reject them.
+ */
+export class DraftManager {
+  /**
+   * The draft copy of the original subtree.
+   */
+  private readonly subtreeRootDraft: object
+
+  /**
+   * The internal `onPatches` disposer.
+   */
+  private readonly disposer: OnPatchesDisposer
+
+  /**
+   * Creates an instance of `DraftManager`.
+   * Do not use directly, use `draftMiddleware` instead.
+   *
+   * @param subtreeRoot Subtree root target object.
+   */
+  public constructor(private readonly subtreeRoot: object) {
+    assertTweakedObject(subtreeRoot, "subtreeRoot")
+    this.subtreeRootDraft = clone(subtreeRoot)
+    if (isRootStore(subtreeRoot)) {
+      registerRootStore(this.subtreeRootDraft)
+    }
+    this.disposer = onPatches(subtreeRoot, patches => applyPatches(this.subtreeRootDraft, patches))
+  }
+
+  /**
+   * Executes `fn` with a draft of `node`. The changes made to the draft in `fn` can be accepted,
+   * i.e. applied to the original subtree, or rejected.
+   *
+   * @typeparam T Object type.
+   * @param node Object for which to obtain a draft copy.
+   * @param fn Function that is caled with a draft copy of `node`. Any changes made to the draft are
+   * applied to the original subtree when `fn` returns `true`. When `fn` returns `false`,
+   * `undefined` or nothing, the changes made to the draft are rejected.
+   */
+  public withDraft<T extends object>(node: T, fn: WithDraftCallback<T>): void {
+    assertTweakedObject(node, "node")
+
+    const path = getParentToChildPath(this.subtreeRoot, node)
+    if (!path) {
+      return
+    }
+
+    const draftNode = resolvePath<T>(this.subtreeRootDraft, path).value
+    if (!draftNode) {
+      return
+    }
+
+    const recorder = patchRecorder(this.subtreeRootDraft)
+    const accept = fn(draftNode)
```

Comment:
try/finally and in the finally block call the disposer (to avoid mem leaks) just in case the userland accept fn throws

Reply:
and maybe run the "non-accept" code in case it throws (to keep the draft consistent with the original in such cases)

Reply:
If the userland accept fn throws and I call the disposer in the finally block but the user catches the exception and recovers properly, won't calling the disposer cause an unexpected state of the `DraftManager` instance? I'd say the user is responsible for disposing of the `DraftManager` instance manually.

Reply:
Sorry, I wasn't specific enough, I meant the recorder disposer 😁

Reply:
Of course, that makes sense. Done in 39e4911f81f180d04b1e727953998bc98cb5f92d.

I decided to handle both `accept` cases in the finally block as I think it's a bit easier to read instead of handling only falsy `accept` in the finally block and truthy `accept` after the finally block. Would you prefer handling only falsy `accept` in the finally block?

Reply:
Looks nice!